### PR TITLE
New field in links (method, headers, body)

### PR DESCRIPTION
**Pull request by emmanuelmathot**
**Related Issue(s):** #1198

**Proposed Changes:**

1. Added `method`, `headers`, `body` fields in `links` objects

**PR Checklist:**

- [X] This PR is made against the dev branch (all proposed changes except releases should be against dev, not master).
- [ ] This PR has **no** breaking changes.
- [X] I have added my changes to the [CHANGELOG](https://github.com/radiantearth/stac-spec/blob/dev/CHANGELOG.md)
      **or** a CHANGELOG entry is not required.
- [X] This PR affects the [STAC API spec](https://github.com/radiantearth/stac-api-spec),
      and I have opened issue https://github.com/radiantearth/stac-api-spec/issues/435 to track the change.

### CODE REVIEW

**Review by m-mohr — COMMENTED** [review]
This looks generally good, but a couple of thoughts (for here and partially also the API):

- headers: Shouldn't this be strictly just be a `Map<string, string>`?
- body: Shouldn't this allow anything? Why is it only an object? Why can't it be a string or an array for example?
- Should we have a central place where we define links so that it's not so repetitive?

**Comment by emmanuelmathot** [comment]
> This looks generally good, but a couple of thoughts (for here and partially also the API):
> 
> * headers: Shouldn't this be strictly just be a `Map<string, string>`?
> * body: Shouldn't this allow anything? Why is it only an object? Why can't it be a string or an array for example?

I used the same spec as per STAC API to be aligned since the initial issue was about being in line with STAC API.

> * Should we have a central place where we define links so that it's not so repetitive?

I moved link object definitions from item, catalog and collection to common metadata page

**Comment by m-mohr** [comment]
> I used the same spec as per STAC API to be aligned since the initial issue was about being in line with STAC API.

Maybe that's something we should actually bring up in the API and change there, too...

File: item-spec/common-metadata.md [hunk 1]
```diff
@@ -207,3 +208,23 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+| rel        | string | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
+| type       | string | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |
+| title      | string | A human readable title to be used in rendered displays of the link.                                                                                                            |
+| method     | string | The HTTP method required to be supported by the target resource. `GET` by default                                                                                              |
+| headers    | object | The HTTP headers required to be sent by the target resource.                                                                                                                   |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:226]: We don't use the object type in the spec, instead we use `Map<string, *>` (applies to headers and body for now)

Comment (m-mohr) [reply item-spec/common-metadata.md:226]: `Map<string, string|array<string>>`

`{a: "b"}`
translates to:
`a: b`

`{a: ["b", "c"]}`
translates to:
```
a: b
a: c
```

This could hopefully also be clarified in the STAC API spec.

File: item-spec/common-metadata.md [hunk 2]
```diff
@@ -221,15 +221,15 @@ with the links section, to describe things like the Catalog an Item is in, relat
...
+| rel        | string         | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
+| type       | string         | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |
+| title      | string         | A human readable title to be used in rendered displays of the link.                                                                                                            |
+| method     | string         | The HTTP method required to be supported by the target resource. `GET` by default                                                                                              |
+| headers    | Map<string, *> | The HTTP headers required to be sent by the target resource.                                                                                                                   |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:231]: I think we still need to explain what we discussed in the last call. Arrays will be expanded to multiple headers.

Ultimately, we could also say:
`Map<string, string | [string]>`

File: item-spec/common-metadata.md [hunk 3]
```diff
@@ -221,15 +221,15 @@ with the links section, to describe things like the Catalog an Item is in, relat
...
+| body       | object         | The HTTP body required to be sent by the target resource.                                                                                                                      |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:232]:
```suggestion
| body       | Map<string, *> | The HTTP body required to be sent by the target resource.                                                                                                                      |
```

Comment (emmanuelmathot) [reply item-spec/common-metadata.md:232]: why the body as Map? body can be anything like the [stac api spec](https://github.com/radiantearth/stac-api-spec/blob/604ade6158de15b8ab068320ca41e25e2bf0e116/core/commons.yaml#L421), no?

Comment (m-mohr) [reply item-spec/common-metadata.md:232]: Right, you are correct. But then it's also not an object. So it should be:

```suggestion
| body       | any            | The HTTP body required to be sent by the target resource.                                                                                                                      |
```

File: item-spec/common-metadata.md (hunk 2 above)
Comment (emmanuelmathot) [reply item-spec/common-metadata.md:231]: explaination added

File: item-spec/common-metadata.md [hunk 4]
```diff
@@ -212,3 +213,23 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+It is allowed to add additional fields such as a `title` and `type`.
+
+| Field Name | Type           | Description                                                                                                                                                                    |
+| ---------- | -------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| href       | string         | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
+| rel        | string         | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
+| type       | string         | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:228]: I guess this shouldn't link to the catalog-spec. 

```suggestion
| type       | string         | Media type) of the referenced entity.                                                                                            |
```

File: item-spec/common-metadata.md (hunk 2 above)
Comment (m-mohr) [reply item-spec/common-metadata.md:231]: Can't find the explanation. Where can it be found?

File: item-spec/json-schema/item.json [hunk 5]
```diff
@@ -203,6 +203,18 @@
...
               "title": "Link title",
               "type": "string"
+            },
+            "method": {
+              "title": "Link method",
+              "type": "string"
+            },
+            "headers": {
+              "title": "Link headers",
+              "type": "object"
```
Comment (m-mohr) [inline item-spec/json-schema/item.json:213]: string or array of strings?!

File: item-spec/json-schema/item.json [hunk 6]
```diff
@@ -203,6 +203,18 @@
...
+            },
+            "body": {
+              "title": "Link body",
+              "type": "object"
```
Comment (m-mohr) [inline item-spec/json-schema/item.json:217]:
```suggestion
              "$comment": "Any type is allowed."
```

File: item-spec/json-schema/item.json [hunk 7]
```diff
@@ -203,6 +203,18 @@
             "title": {
               "title": "Link title",
               "type": "string"
+            },
+            "method": {
+              "title": "Link method",
+              "type": "string"
```
Comment (m-mohr) [inline item-spec/json-schema/item.json:209]: Should this be an enum?

File: item-spec/common-metadata.md [hunk 8]
```diff
@@ -212,3 +213,29 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+
+| Field Name | Type                             | Description                                                                                                                                                                    |
+| ---------- | -------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
+| href       | string                           | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
+| rel        | string                           | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
+| type       | string                           | [Media type](../catalog-spec/catalog-spec.md#media-types) of the referenced entity.                                                                                            |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:228]:
```suggestion
| type       | string                           | Media type of the referenced entity.                                                                                            |
```

File: item-spec/common-metadata.md [hunk 9]
```diff
@@ -212,3 +213,29 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+| title      | string                           | A human readable title to be used in rendered displays of the link.                                                                                                            |
+| method     | string                           | The HTTP method required to be supported by the target resource. `GET` by default                                                                                              |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:230]:
```suggestion
| method     | string                           | The HTTP method that shall be used for the request to the target resource. `GET` by default                                                                                              |
```

File: item-spec/common-metadata.md [hunk 10]
```diff
@@ -212,3 +213,29 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+| headers    | Map<string, string \| \[string]> | The HTTP headers required to be sent by the target resource.                                                                                                                   |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:231]:
```suggestion
| headers    | Map<string, string \| \[string]> | The HTTP headers to be sent for the request to the target resource.                                                                                                                    |
```

File: item-spec/common-metadata.md [hunk 11]
```diff
@@ -212,3 +213,29 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+| body       | any                              | The HTTP body required to be sent by the target resource.                                                                                                                      |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:232]:
```suggestion
| body       | any                              | The HTTP body to be sent to the target resource.                                                                                                                      |
```

File: item-spec/common-metadata.md [hunk 12]
```diff
@@ -212,3 +213,29 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+
+#### HTTP headers
+
+The field `headers` allows to describe a dictionary of HTTP headers that are required to be sent by the client.
+The keys of the dictionary are the header names, and the values are either a single string or an array of strings.
+In case of an array, the header is expected to be sent multiple times with the different values.
+
+For a full discussion of the situations where relative and absolute links are recommended see the
+['Use of links'](../best-practices.md#use-of-links) section of the STAC best practices.
```
Comment (m-mohr) [inline item-spec/common-metadata.md:241]: This seems unrelated to HTTP headers. Should this be in the other section below the table?

File: item-spec/common-metadata.md [hunk 13]
```diff
@@ -212,3 +213,29 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
 PlanetScope Ortho Tile Product has an `gsd` of 3.7 (or 4 if rounding), even though the pixel size of the images is 3.125.
 For example, one might choose for WorldView-2 the Multispectral 20° off-nadir value of 2.07
 and for WorldView-3 the Multispectral 20° off-nadir value of 1.38.
+
+### Link Object
+
+This object describes a relationship with another entity. Data providers are advised to be liberal
+with the links section, to describe things like the Catalog an Item is in, related Items, parent or
+child Items (modeled in different ways, like an 'acquisition' or derived data).
+It is allowed to add additional fields such as a `title` and `type`.
```
Comment (gadomski) [inline item-spec/common-metadata.md:222]: Is this sentence correct? `title` and `type` are both fields described in the table below.

File: item-spec/json-schema/item.json (hunk 7 above)
Comment (m-mohr) [reply item-spec/json-schema/item.json:209]: @emmanuelmathot I think we should provide an enum here so that it's (1) clear which ones are allowed and (2) whether they are uppercase or lowercase (I think it's all uppercase).

File: item-spec/common-metadata.md [hunk 14]
```diff
@@ -212,3 +213,28 @@ optical and short-wave IR bands are all 30 meters, but the panchromatic band is
...
+| href       | string                           | **REQUIRED.** The actual link in the format of an URL. Relative and absolute links are both allowed. [Trailing slashes are significant.](../best-practices.md#consistent-uris) |
+| rel        | string                           | **REQUIRED.** Relationship between the current document and the linked document. See chapter "Relation types" for more information.                                            |
+| type       | string                           | Media type of the referenced entity.                                                                                                                                           |
+| title      | string                           | A human readable title to be used in rendered displays of the link.                                                                                                            |
+| method     | string                           | The HTTP method that shall be used for the request to the target resource. `GET` by default                                                                                    |
```
Comment (m-mohr) [inline item-spec/common-metadata.md:229]:
```suggestion
| method     | string                           | The HTTP method that shall be used for the request to the target resource, in uppercase. `GET` by default                                                                      |
```

**Comment by m-mohr** [comment]
Looks solid, I just proposed to additional clarifications around the method. Happy to approve afterwards.

**Review by m-mohr — APPROVED** [review]

**Review by gadomski — APPROVED** [review]
